nova/src/relaxed_r1cs.rs

```diff
@@ -1,113 +1,86 @@
 mod instance;
 mod witness;
 
-use crate::hash::MimcRO;
-pub(crate) use instance::RelaxedR1csInstance;
-pub(crate) use witness::RelaxedR1csWitness;
+use crate::PedersenCommitment;
+pub(crate) use instance::{R1csInstance, RelaxedR1csInstance};
+pub(crate) use witness::{R1csWitness, RelaxedR1csWitness};
 use zkstd::circuit::prelude::{CircuitDriver, R1cs};
+use zkstd::common::Ring;
 use zkstd::matrix::{DenseVectors, SparseMatrix};
 
 #[derive(Clone, Debug)]
-pub struct RelaxedR1cs<C: CircuitDriver> {
-    // 1. Structure S
-    // a, b and c matrices and matrix size
+pub struct R1csShape<C: CircuitDriver> {
     m: usize,
+    instance_length: usize,
+    witness_length: usize,
     a: SparseMatrix<C::Scalar>,
     b: SparseMatrix<C::Scalar>,
     c: SparseMatrix<C::Scalar>,
-
-    // 2. Instance
-    // r1cs instance includes public inputs, outputs and scalar
-    pub(crate) instance: RelaxedR1csInstance<C>,
-
-    // 3. Witness
-    // r1cs witness includes private inputs, intermediate value and error vector
-    pub(crate) witness: RelaxedR1csWitness<C>,
 }
 
-impl<C: CircuitDriver> RelaxedR1cs<C> {
-    pub fn new(r1cs: R1cs<C>) -> Self {
-        let m = r1cs.m();
-        let (a, b, c) = r1cs.matrices();
-        let x = DenseVectors::new(r1cs.x());
-        let w = DenseVectors::new(r1cs.w());
-
-        let instance = RelaxedR1csInstance::new(x);
-        let witness = RelaxedR1csWitness::new(w, m);
+pub(crate) fn r1cs_instance_and_witness<C: CircuitDriver>(
```

Comment:
Get instance/witness values from R1cs that should match the provided shape.
R1csWitness and R1csInstance can be relaxed after if necessary.